modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java

```diff
@@ -3301,21 +3325,28 @@ private Set<TableImpl> zoneTablesRawSet(int zoneId) {
      * @param table Table to add.
      */
     private void addTableToZone(int zoneId, TableImpl table) {
-        StampedLock rwLock = tablesPerZoneLocks.computeIfAbsent(zoneId, id -> new StampedLock());
+        NaiveAsyncReadWriteLock zoneLock = tablesPerZoneLocks.computeIfAbsent(zoneId, id -> new NaiveAsyncReadWriteLock());
+
+        CompletableFuture<Long> writeLockAcquisitionFuture = zoneLock.writeLock();
 
-        long stamp = rwLock.writeLock();
         try {
-            tablesPerZone.compute(zoneId, (id, tables) -> {
-                if (tables == null) {
-                    tables = new HashSet<>();
-                }
+            writeLockAcquisitionFuture.thenAccept(stamp -> {
+                tablesPerZone.compute(zoneId, (id, tables) -> {
+                    if (tables == null) {
+                        tables = new HashSet<>();
+                    }
 
-                tables.add(table);
+                    tables.add(table);
 
-                return tables;
-            });
-        } finally {
-            rwLock.unlockWrite(stamp);
+                    return tables;
+                });
+
+                zoneLock.unlockWrite(stamp);
+            }).join();
```

Comment:
Same note about blocking the calling thread. Do we really need it? If yes, this looks very suspicious, it smells like a possible deadlock